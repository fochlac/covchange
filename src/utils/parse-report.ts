import { parse } from 'fast-xml-parser'

const config = {
	ignoreAttributes: false,
	attrNodeName: 'attr',
	attributeNamePrefix: '',
	parseAttributeValue: true,
}

/**
 * Function to parse a raw clover coverage report in xml format
 *
 * @param {string} rawXML coverage report as utf8 string
 * @returns {Core.Report} parsed coverage report
 */
export function createReportFromXML(rawXML: string): Core.Report {
	const rawProject = parse(rawXML, config).coverage.project

	const folders = rawProject.metrics.package.reduce(reduceFolder, {})
	const files = Object.values(folders).reduce(reduceFiles, {})

	return {
		timestamp: rawProject.attr.timestamp,
		metrics: parseMetrics(rawProject.metrics.attr),
		folders,
		files,
	}
}

function reduceFolder(folders, folder) {
	folders[folder.attr.name] = {
		name: folder.attr.name,
		metrics: parseMetrics(folder.metrics.attr),
		files: parseFile(folder.file),
	}
	return folders
}

function reduceFiles(files, folder) {
	folder.files.forEach(file => {
		files[`${folder.name.split('.').join('/')}/${file.name}`] = file
	})
	return files
}

function parseFile(file) {
	const rawFiles = file && file.length ? file : file ? [file] : []

	return rawFiles.map(file => ({
		name: file.attr.name,
		metrics: parseMetrics(file.metrics.attr),
	}))
}

function parseMetrics({ statements, coveredstatements, conditionals, coveredconditionals }) {
	return {
		statements,
		conditionals,
		coveredStatements: coveredstatements,
		coveredConditionals: coveredconditionals,
		statementCov: (statements && Math.round((coveredstatements / statements) * 10000)) / 100 || 0,
		conditionalCov: (conditionals && Math.round((coveredconditionals / conditionals) * 10000)) / 100 || 0,
	}
}
