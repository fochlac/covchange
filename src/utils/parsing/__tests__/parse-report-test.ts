import { createReportFromXML } from '../parse-report'
import { reportRaw, reportRaw3 } from '../../../__tests__/data/reports'

describe('createReportFromXML', () => {
	it('should properly parse an xml file', () => {
		expect(createReportFromXML(reportRaw)).toMatchSnapshot()
		expect(createReportFromXML(reportRaw3)).toMatchSnapshot()
	})
})
