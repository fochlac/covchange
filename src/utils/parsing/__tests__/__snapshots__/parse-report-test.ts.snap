// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReportFromXML should properly parse an xml file 1`] = `
Object {
  "files": Object {
    "orders/app/components/order-tours-tab/order-tours-tab.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 35,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 34,
      },
      "name": "order-tours-tab.js",
    },
    "orders/app/components/order-tours-tab/order-tours-table-config.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 26,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 8,
      },
      "name": "order-tours-table-config.js",
    },
    "orders/app/components/order-tours-tab/order-tours-table-renderer.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 35,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 35,
      },
      "name": "order-tours-table-renderer.js",
    },
    "orders/app/components/order-tours-tab/order-tours-table.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 41,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 27,
      },
      "name": "order-tours-table.js",
    },
    "orders/app/components/orders-filter-bar.js": Object {
      "metrics": Object {
        "conditionalCov": 31.82,
        "conditionals": 22,
        "coveredConditionals": 7,
        "coveredStatements": 56,
        "statementCov": 83.58,
        "statements": 67,
      },
      "name": "orders-filter-bar.js",
    },
    "orders/app/components/orders-subviews.js": Object {
      "metrics": Object {
        "conditionalCov": 100,
        "conditionals": 17,
        "coveredConditionals": 17,
        "coveredStatements": 17,
        "statementCov": 80.95,
        "statements": 21,
      },
      "name": "orders-subviews.js",
    },
    "orders/app/components/orders-tab/order-type-filter-dropdown.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 0,
        "coveredConditionals": 0,
        "coveredStatements": 6,
        "statementCov": 100,
        "statements": 6,
      },
      "name": "order-type-filter-dropdown.js",
    },
    "orders/app/components/orders-tab/orders-date-picker.js": Object {
      "metrics": Object {
        "conditionalCov": 50,
        "conditionals": 2,
        "coveredConditionals": 1,
        "coveredStatements": 5,
        "statementCov": 100,
        "statements": 5,
      },
      "name": "orders-date-picker.js",
    },
    "orders/app/components/orders-tab/orders-tab.js": Object {
      "metrics": Object {
        "conditionalCov": 83.33,
        "conditionals": 6,
        "coveredConditionals": 5,
        "coveredStatements": 17,
        "statementCov": 65.38,
        "statements": 26,
      },
      "name": "orders-tab.js",
    },
    "orders/app/components/orders-view.js": Object {
      "metrics": Object {
        "conditionalCov": 100,
        "conditionals": 17,
        "coveredConditionals": 17,
        "coveredStatements": 47,
        "statementCov": 100,
        "statements": 47,
      },
      "name": "orders-view.js",
    },
    "orders/app/index.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 12,
      },
      "name": "index.js",
    },
    "orders/module.js": Object {
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 1,
      },
      "name": "module.js",
    },
  },
  "folders": Object {
    "orders": Object {
      "files": Array [
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 4,
            "coveredConditionals": 0,
            "coveredStatements": 0,
            "statementCov": 0,
            "statements": 1,
          },
          "name": "module.js",
        },
      ],
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 1,
      },
      "name": "orders",
    },
    "orders.app": Object {
      "files": Array [
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 4,
            "coveredConditionals": 0,
            "coveredStatements": 0,
            "statementCov": 0,
            "statements": 12,
          },
          "name": "index.js",
        },
      ],
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 12,
      },
      "name": "orders.app",
    },
    "orders.app.components": Object {
      "files": Array [
        Object {
          "metrics": Object {
            "conditionalCov": 31.82,
            "conditionals": 22,
            "coveredConditionals": 7,
            "coveredStatements": 56,
            "statementCov": 83.58,
            "statements": 67,
          },
          "name": "orders-filter-bar.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 100,
            "conditionals": 17,
            "coveredConditionals": 17,
            "coveredStatements": 17,
            "statementCov": 80.95,
            "statements": 21,
          },
          "name": "orders-subviews.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 100,
            "conditionals": 17,
            "coveredConditionals": 17,
            "coveredStatements": 47,
            "statementCov": 100,
            "statements": 47,
          },
          "name": "orders-view.js",
        },
      ],
      "metrics": Object {
        "conditionalCov": 73.21,
        "conditionals": 56,
        "coveredConditionals": 41,
        "coveredStatements": 120,
        "statementCov": 88.89,
        "statements": 135,
      },
      "name": "orders.app.components",
    },
    "orders.app.components.order-tours-tab": Object {
      "files": Array [
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 35,
            "coveredConditionals": 0,
            "coveredStatements": 0,
            "statementCov": 0,
            "statements": 34,
          },
          "name": "order-tours-tab.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 26,
            "coveredConditionals": 0,
            "coveredStatements": 0,
            "statementCov": 0,
            "statements": 8,
          },
          "name": "order-tours-table-config.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 35,
            "coveredConditionals": 0,
            "coveredStatements": 0,
            "statementCov": 0,
            "statements": 35,
          },
          "name": "order-tours-table-renderer.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 41,
            "coveredConditionals": 0,
            "coveredStatements": 0,
            "statementCov": 0,
            "statements": 27,
          },
          "name": "order-tours-table.js",
        },
      ],
      "metrics": Object {
        "conditionalCov": 0,
        "conditionals": 137,
        "coveredConditionals": 0,
        "coveredStatements": 0,
        "statementCov": 0,
        "statements": 104,
      },
      "name": "orders.app.components.order-tours-tab",
    },
    "orders.app.components.orders-tab": Object {
      "files": Array [
        Object {
          "metrics": Object {
            "conditionalCov": 0,
            "conditionals": 0,
            "coveredConditionals": 0,
            "coveredStatements": 6,
            "statementCov": 100,
            "statements": 6,
          },
          "name": "order-type-filter-dropdown.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 50,
            "conditionals": 2,
            "coveredConditionals": 1,
            "coveredStatements": 5,
            "statementCov": 100,
            "statements": 5,
          },
          "name": "orders-date-picker.js",
        },
        Object {
          "metrics": Object {
            "conditionalCov": 83.33,
            "conditionals": 6,
            "coveredConditionals": 5,
            "coveredStatements": 17,
            "statementCov": 65.38,
            "statements": 26,
          },
          "name": "orders-tab.js",
        },
      ],
      "metrics": Object {
        "conditionalCov": 75,
        "conditionals": 8,
        "coveredConditionals": 6,
        "coveredStatements": 28,
        "statementCov": 75.68,
        "statements": 37,
      },
      "name": "orders.app.components.orders-tab",
    },
  },
  "metrics": Object {
    "conditionalCov": 72.85,
    "conditionals": 2479,
    "coveredConditionals": 1806,
    "coveredStatements": 3428,
    "statementCov": 84.81,
    "statements": 4042,
  },
  "timestamp": 1543960595393,
}
`;
